### PR TITLE
Blaze: Add picker view for target languages

### LINKED ISSUE #11630: Edit language picker
image

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #11630 

## Description

This PR adds a new screen for selecting target languages for a creating a new Blaze campaign. Changes include:
- Updated all Blaze target option models to conform to `Hashable` and `Identifiable`. This is necessary to display the models on a SwiftUI list.
- Renamed `SelectionList` to `SingleSelectionList`. For simplicity, I created a separate reusable view `MultiSelectionList` for use in the Blaze campaign creation flow.
- Created `BlazeTargetLanguagePickerView` and view model to sync and display languages for picking.
- Integrated the language picker view to the Blaze campaign creation form.

## Testing instructions

- Log in to a store with at least one existing product and is eligible for Blaze.
- Navigate to the Products tab and select any public product.
- On the product form, select Promote with Blaze.
- Notice that the campaign creation form is presented. Select the language field.
- Notice that a new screen is displayed for selecting target languages. Confirm that the selection and filtering work as expected.
- Select Cancel, the view should be dismissed without any changes to the campaign creation form.
- Select the language field again, this time select a different option and tap Save. The language field in the creation form should be updated correctly.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/5533851/a4a12605-5e74-4c03-a0a5-fc3d4876d608

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift [hunk 1]
```diff
@@ -22,16 +22,19 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
...
 
     var targetOptions: BlazeTargetOptions? {
         guard locations != nil || languages != nil || devices != nil || pageTopics != nil else {
             return nil
         }
-        return BlazeTargetOptions(locations: locations, languages: languages, devices: devices, pageTopics: pageTopics)
+        return BlazeTargetOptions(locations: locations?.map { $0.id },
+                                  languages: languages?.map { $0.id },
+                                  devices: devices?.map { $0.id },
+                                  pageTopics: pageTopics?.map { $0.id })
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:37]: nit: Consider unwrapping these values in the above `guard` statement. 

```swift
        guard let locations, let languages, let devices, let pageTopics else {
            return nil
        }
        return BlazeTargetOptions(locations: locations.map { $0.id },
                                  languages: languages.map { $0.id },
                                  devices: devices.map { $0.id },
                                  pageTopics: pageTopics.map { $0.id })
```

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SingleSelectionList.swift [hunk 2]
```diff
@@ -77,7 +78,7 @@ struct SelectionList<T: Hashable>: View {
 
 struct SelectionList_Previews: PreviewProvider {
     static var previews: some View {
-        SelectionList(title: "Lunch",
+        SingleSelectionList(title: "Lunch",
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SingleSelectionList.swift:81]: Super nit: Indentation needs to be updated.

File: WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeTargetLanguagePickerViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        // When
+        viewModel.searchQuery = ""
+
+        // Then
+        waitUntil {
+            viewModel.languages == [english, vietnamese]
+        }
+    }
+
+    func test_languages_filters_matching_languages_if_searchQuery_is_empty() {
```
Comment (selanthiraiyan) [inline WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazeTargetLanguagePickerViewModelTests.swift]:
```suggestion
    func test_languages_filters_matching_languages_if_searchQuery_is_not_empty() {
```

File: WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsTimeRangeCard.swift [hunk 4]
```diff
@@ -37,7 +37,7 @@ struct AnalyticsTimeRangeCard: View {
     var body: some View {
         createTimeRangeContent()
             .sheet(isPresented: $showTimeRangeSelectionView) {
-                SelectionList(title: Localization.timeRangeSelectionTitle,
+                SingleSelectionList(title: Localization.timeRangeSelectionTitle,
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsTimeRangeCard.swift:40]: Super nit: Indentation fix needed. This applies to other places where we have replaced `SelectionList` with `SingleSelectionList`.

File: WooCommerce/Classes/ViewRelated/Blaze/TargetLanguages/BlazeTargetLanguagePickerViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import Foundation
+import Yosemite
+import protocol Storage.StorageManagerType
+
+/// View model for `BlazeTargetLanguagePickerView`
+final class BlazeTargetLanguagePickerViewModel: ObservableObject {
+
+    @Published var searchQuery: String = ""
+    @Published private(set) var languages: [BlazeTargetLanguage] = []
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Blaze/TargetLanguages/BlazeTargetLanguagePickerViewModel.swift:10]: nit: Consider adding a comment to explain that holds the `languages` after filtering based on search query.

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/MultiSelectionList.swift [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                            Image(uiImage: .checkmarkStyledImage)
+                                .renderedIf(selectedItems.isEmpty)
+                        }
+                        .contentShape(Rectangle())
+                        .onTapGesture {
+                            if selectedItems.isEmpty {
+                                selectedItems = Set(contents)
+                            } else {
+                                selectedItems = []
+                            }
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/MultiSelectionList.swift:75]: AllSelection

Should we select all items when `selectedItems.isEmpty` is empty? 

---

https://github.com/woocommerce/woocommerce-ios/assets/524475/5b146fb3-9097-4d71-a04c-7a2e335244cd


In the above scenario, I have selected all the available languages, but I don't see "All" being displayed in the creation form. 

I am not sure what is the expected behaviour here.

**Review by selanthiraiyan — APPROVED** [review]
Thanks for working on adding a `MultiSelectionList`, Huong! 

This looks great. I left a few non-blocking nits. 🚀

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:37]: This is not correct. We don't expect every option to be available to create a non-nil option object. I'll keep my change as-is.

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SingleSelectionList.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SingleSelectionList.swift:81]: Updated in 9137b6fd0a32b44501ee83f05bdbeeef06dec8fa

File: WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsTimeRangeCard.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsTimeRangeCard.swift:40]: Updated in 9137b6fd0a32b44501ee83f05bdbeeef06dec8fa

File: WooCommerce/Classes/ViewRelated/Blaze/TargetLanguages/BlazeTargetLanguagePickerViewModel.swift (hunk 5 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Blaze/TargetLanguages/BlazeTargetLanguagePickerViewModel.swift:10]: Added in 2533ea0ca946a5582f30cb0d89dcca28e8d3284f.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/MultiSelectionList.swift (hunk 6 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/MultiSelectionList.swift:75]: Good point! I updated the logic in be41f3e8ba21826521651a120994e41a8c830773 - the All row should be selected if all other rows are and not when any of the other rows are unselected.

**Comment by itsmeichigo** [comment]
Thank you @selanthiraiyan for the reviews! I addressed the issues mentioned in your comments - will merge now to continue with other tasks. Let me know if you have any other comments, I'll handle them in the next PRs.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**